File: multi-agent-ui/src/api/story/video.ts
import request, { type ApiResult } from "@/utils/request";
import type { PageResult } from "@/api/rag";
import { STORY_PROJECT_URL, STORY_VIDEO_URL } from "./index";

// ============================================================
// 类型定义（对齐后端 routers/story 契约，字段一律 snake_case）
// ============================================================

/** 视频成品 VO（单镜头片段） */
export interface StoryVideoVO {
  id: string;
  project_id: string;
  keyframe_id: string | null;
  script_id: string | null;
  export_package_id: string | null;
  title: string | null;
  episode_no: number | null;
  video_file: string;
  cover_file: string | null;
  duration_ms: number | null;
  file_size: number | null;
  width: number | null;
  height: number | null;
  target_platform: string | null;
  external_task_id: string | null;
  status: string;
  remark: string | null;
  created_at: string;
  updated_at: string;
}

/** 视频登记附属字段（上传时可选填） */
export interface StoryVideoRegisterFields {
  title?: string;
  episode_no?: number;
  keyframe_id?: string;
  script_id?: string;
  export_package_id?: string;
  target_platform?: string;
  external_task_id?: string;
  remark?: string;
}

/** 编辑视频请求体（溯源字段不可改） */
export interface StoryVideoUpdatePayload {
  title?: string | null;
  episode_no?: number | null;
  target_platform?: string | null;
  external_task_id?: string | null;
  remark?: string | null;
}

// ============================================================
// API
// ============================================================

export const videoApi = {
  /** 上传视频片段并登记（服务端抽首帧封面） */
  upload(projectId: string, file: File, fields: StoryVideoRegisterFields = {}) {
    const form = new FormData();
    form.append("file", file);
    if (fields.title) form.append("title", fields.title);
    if (fields.episode_no !== undefined) form.append("episode_no", String(fields.episode_no));
    if (fields.keyframe_id) form.append("keyframe_id", fields.keyframe_id);
    if (fields.script_id) form.append("script_id", fields.script_id);
    if (fields.export_package_id) form.append("export_package_id", fields.export_package_id);
    if (fields.target_platform) form.append("target_platform", fields.target_platform);
    if (fields.external_task_id) form.append("external_task_id", fields.external_task_id);
    if (fields.remark) form.append("remark", fields.remark);
    return request.post<StoryVideoVO, ApiResult<StoryVideoVO>>(
      STORY_PROJECT_URL.videos(projectId),
      form,
      { headers: { "Content-Type": "multipart/form-data" }, timeout: 120000 }
    );
  },
  /** 项目视频列表（可选按关键帧过滤） */
  list(projectId: string, params: { page?: number; size?: number; keyframe_id?: string } = {}) {
    return request.get<PageResult<StoryVideoVO>, ApiResult<PageResult<StoryVideoVO>>>(
      STORY_PROJECT_URL.videos(projectId),
      { params }
    );
  },
  /** 编辑视频登记字段 */
  update(id: string, payload: StoryVideoUpdatePayload) {
    return request.put<StoryVideoVO, ApiResult<StoryVideoVO>>(
      STORY_VIDEO_URL.byId(id),
      payload
    );
  },
  /** 删除视频登记 */
  remove(id: string) {
    return request.delete<null, ApiResult<null>>(STORY_VIDEO_URL.byId(id));
  },
  /** 手动上传封面 */
  uploadCover(id: string, file: File) {
    const form = new FormData();
    form.append("file", file);
    return request.post<StoryVideoVO, ApiResult<StoryVideoVO>>(
      STORY_VIDEO_URL.cover(id),
      form,
      { headers: { "Content-Type": "multipart/form-data" } }
    );
  },
  /** 把视频封面设为项目封面 */
  setProjectCover(id: string) {
    return request.post<unknown, ApiResult<unknown>>(STORY_VIDEO_URL.projectCover(id));
  },
};

File: entity/story/video.py
"""
剧本模块视频成品域 ORM 模型（映射 story.story_videos）。

视频成品为外部视频网站生成的单镜头片段，由用户手动上传回收登记；
系统不做视频生成，也不建模整集装配。
"""

import uuid
from datetime import datetime

from sqlalchemy import BigInteger, Integer, String, Text, func, text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from entity.base import Base
from utils.date_format import format_datetime
from utils.id import format_id


class StoryVideo(Base):
    """
    视频成品 ORM 模型（映射 story.story_videos）。

    keyframe_id 为主溯源轴（一帧可多条重抽），script_id / export_package_id
    并列溯源；episode_no 语义放宽为可选分组/排序号；
    cover_file 默认抽视频首帧，失败留空可手动上传。
    """

    __tablename__ = "story_videos"
    __table_args__ = {"schema": "story"}

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("uuidv7()"),
    )

    project_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    keyframe_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )
    script_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )
    export_package_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )
    title: Mapped[str | None] = mapped_column(Text, nullable=True)
    episode_no: Mapped[int | None] = mapped_column(Integer, nullable=True)
    video_file: Mapped[str] = mapped_column(String(500), nullable=False)
    cover_file: Mapped[str | None] = mapped_column(String(500), nullable=True)
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    file_size: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    width: Mapped[int | None] = mapped_column(Integer, nullable=True)
    height: Mapped[int | None] = mapped_column(Integer, nullable=True)
    target_platform: Mapped[str | None] = mapped_column(String(100), nullable=True)
    external_task_id: Mapped[str | None] = mapped_column(String(200), nullable=True)
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, server_default=text("'done'")
    )
    remark: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        nullable=False, server_default=func.now()
    )

    def to_dict(self) -> dict:
        """转为可 JSON 序列化的普通字典，供统一响应回写。"""
        return {
            "id": format_id(self.id),
            "project_id": format_id(self.project_id),
            "keyframe_id": format_id(self.keyframe_id),
            "script_id": format_id(self.script_id),
            "export_package_id": format_id(self.export_package_id),
            "title": self.title,
            "episode_no": self.episode_no,
            "video_file": self.video_file,
            "cover_file": self.cover_file,
            "duration_ms": self.duration_ms,
            "file_size": self.file_size,
            "width": self.width,
            "height": self.height,
            "target_platform": self.target_platform,
            "external_task_id": self.external_task_id,
            "status": self.status,
            "remark": self.remark,
            "created_at": format_datetime(self.created_at),
            "updated_at": format_datetime(self.updated_at),
        }
